### DISCUSSION TITLE
Building WiX with VS2022

### DISCUSSION THREAD

**Discussion by MarkStega** [wixtoolset/issues#6662] [Questions] [answered]
I am trying to build WiX using VS2022 17.0.2; I do not have previous VS versions installed on my development system.

I have the appropriate workloads installed & added the V141&v142 c++ build tools for all 3 

Using the head commit from the develop branch I receive the following warnings:

```
C:\Solutions\TemporarySolutions\wix4\src\internal\WixBuildTools.TestSupport\WixBuildTools.TestSupport.csproj : warning NU1603: WixBuildTools.
TestSupport depends on WixToolset.Dtf.Compression (>= 4.0.0-preview.1-build.88) but WixToolset.Dtf.Compression 4.0.0-preview.1-build.88 was n
ot found. An approximate best match of WixToolset.Dtf.Compression 4.0.0-preview.1-build.89 was resolved.
C:\Solutions\TemporarySolutions\wix4\src\internal\WixBuildTools.TestSupport\WixBuildTools.TestSupport.csproj : warning NU1603: WixBuildTools.
TestSupport depends on WixToolset.Dtf.Compression.Cab (>= 4.0.0-preview.1-build.88) but WixToolset.Dtf.Compression.Cab 4.0.0-preview.1-build.
88 was not found. An approximate best match of WixToolset.Dtf.Compression.Cab 4.0.0-preview.1-build.89 was resolved.
C:\Solutions\TemporarySolutions\wix4\src\internal\WixBuildTools.TestSupport\WixBuildTools.TestSupport.csproj : warning NU1603: WixBuildTools.
TestSupport depends on WixToolset.Dtf.WindowsInstaller (>= 4.0.0-preview.1-build.88) but WixToolset.Dtf.WindowsInstaller 4.0.0-preview.1-buil
d.88 was not found. An approximate best match of WixToolset.Dtf.WindowsInstaller 4.0.0-preview.1-build.89 was resolved.
  WixBuildTools.TestSupport -> C:\Solutions\TemporarySolutions\wix4\build\internal\Release\netstandard2.0\WixBuildTools.TestSupport.dll
C:\Solutions\TemporarySolutions\wix4\src\internal\WixBuildTools.TestSupport\WixBuildTools.TestSupport.csproj : warning NU1603: WixBuildTools.
TestSupport depends on WixToolset.Dtf.Compression (>= 4.0.0-preview.1-build.88) but WixToolset.Dtf.Compression 4.0.0-preview.1-build.88 was n
ot found. An approximate best match of WixToolset.Dtf.Compression 4.0.0-preview.1-build.89 was resolved.
C:\Solutions\TemporarySolutions\wix4\src\internal\WixBuildTools.TestSupport\WixBuildTools.TestSupport.csproj : warning NU1603: WixBuildTools.
TestSupport depends on WixToolset.Dtf.Compression.Cab (>= 4.0.0-preview.1-build.88) but WixToolset.Dtf.Compression.Cab 4.0.0-preview.1-build.
88 was not found. An approximate best match of WixToolset.Dtf.Compression.Cab 4.0.0-preview.1-build.89 was resolved.
C:\Solutions\TemporarySolutions\wix4\src\internal\WixBuildTools.TestSupport\WixBuildTools.TestSupport.csproj : warning NU1603: WixBuildTools.
TestSupport depends on WixToolset.Dtf.WindowsInstaller (>= 4.0.0-preview.1-build.88) but WixToolset.Dtf.WindowsInstaller 4.0.0-preview.1-buil
d.88 was not found. An approximate best match of WixToolset.Dtf.WindowsInstaller 4.0.0-preview.1-build.89 was resolved.
  WixBuildTools.TestSupport -> C:\Solutions\TemporarySolutions\wix4\build\internal\Release\net461\WixBuildTools.TestSupport.dll
C:\Solutions\TemporarySolutions\wix4\src\internal\WixBuildTools.TestSupport\WixBuildTools.TestSupport.csproj : warning NU1603: WixBuildTools.
TestSupport depends on WixToolset.Dtf.Compression (>= 4.0.0-preview.1-build.88) but WixToolset.Dtf.Compression 4.0.0-preview.1-build.88 was n
ot found. An approximate best match of WixToolset.Dtf.Compression 4.0.0-preview.1-build.89 was resolved.
C:\Solutions\TemporarySolutions\wix4\src\internal\WixBuildTools.TestSupport\WixBuildTools.TestSupport.csproj : warning NU1603: WixBuildTools.
TestSupport depends on WixToolset.Dtf.Compression.Cab (>= 4.0.0-preview.1-build.88) but WixToolset.Dtf.Compression.Cab 4.0.0-preview.1-build.
88 was not found. An approximate best match of WixToolset.Dtf.Compression.Cab 4.0.0-preview.1-build.89 was resolved.
C:\Solutions\TemporarySolutions\wix4\src\internal\WixBuildTools.TestSupport\WixBuildTools.TestSupport.csproj : warning NU1603: WixBuildTools.
TestSupport depends on WixToolset.Dtf.WindowsInstaller (>= 4.0.0-preview.1-build.88) but WixToolset.Dtf.WindowsInstaller 4.0.0-preview.1-buil
d.88 was not found. An approximate best match of WixToolset.Dtf.WindowsInstaller 4.0.0-preview.1-build.89 was resolved.
  WixBuildTools.TestSupport -> C:\Solutions\TemporarySolutions\wix4\build\internal\Release\net472\WixBuildTools.TestSupport.dll
C:\Solutions\TemporarySolutions\wix4\src\internal\WixBuildTools.TestSupport\WixBuildTools.TestSupport.csproj : warning NU1603: WixBuildTools.
TestSupport depends on WixToolset.Dtf.Compression (>= 4.0.0-preview.1-build.88) but WixToolset.Dtf.Compression 4.0.0-preview.1-build.88 was n
ot found. An approximate best match of WixToolset.Dtf.Compression 4.0.0-preview.1-build.89 was resolved.
C:\Solutions\TemporarySolutions\wix4\src\internal\WixBuildTools.TestSupport\WixBuildTools.TestSupport.csproj : warning NU1603: WixBuildTools.
TestSupport depends on WixToolset.Dtf.Compression.Cab (>= 4.0.0-preview.1-build.88) but WixToolset.Dtf.Compression.Cab 4.0.0-preview.1-build.
88 was not found. An approximate best match of WixToolset.Dtf.Compression.Cab 4.0.0-preview.1-build.89 was resolved.
C:\Solutions\TemporarySolutions\wix4\src\internal\WixBuildTools.TestSupport\WixBuildTools.TestSupport.csproj : warning NU1603: WixBuildTools.
TestSupport depends on WixToolset.Dtf.WindowsInstaller (>= 4.0.0-preview.1-build.88) but WixToolset.Dtf.WindowsInstaller 4.0.0-preview.1-buil
d.88 was not found. An approximate best match of WixToolset.Dtf.WindowsInstaller 4.0.0-preview.1-build.89 was resolved.
C:\Solutions\TemporarySolutions\wix4\src\internal\WixBuildTools.TestSupport\WixBuildTools.TestSupport.csproj : warning NU1603: WixBuildTools.
TestSupport depends on WixToolset.Dtf.Compression (>= 4.0.0-preview.1-build.88) but WixToolset.Dtf.Compression 4.0.0-preview.1-build.88 was n
ot found. An approximate best match of WixToolset.Dtf.Compression 4.0.0-preview.1-build.89 was resolved.
C:\Solutions\TemporarySolutions\wix4\src\internal\WixBuildTools.TestSupport\WixBuildTools.TestSupport.csproj : warning NU1603: WixBuildTools.
TestSupport depends on WixToolset.Dtf.Compression.Cab (>= 4.0.0-preview.1-build.88) but WixToolset.Dtf.Compression.Cab 4.0.0-preview.1-build.
88 was not found. An approximate best match of WixToolset.Dtf.Compression.Cab 4.0.0-preview.1-build.89 was resolved.
C:\Solutions\TemporarySolutions\wix4\src\internal\WixBuildTools.TestSupport\WixBuildTools.TestSupport.csproj : warning NU1603: WixBuildTools.
TestSupport depends on WixToolset.Dtf.WindowsInstaller (>= 4.0.0-preview.1-build.88) but WixToolset.Dtf.WindowsInstaller 4.0.0-preview.1-buil
d.88 was not found. An approximate best match of WixToolset.Dtf.WindowsInstaller 4.0.0-preview.1-build.89 was resolved.
C:\Solutions\TemporarySolutions\wix4\src\internal\WixBuildTools.TestSupport\WixBuildTools.TestSupport.csproj : warning NU1603: WixBuildTools.
TestSupport depends on WixToolset.Dtf.Compression (>= 4.0.0-preview.1-build.88) but WixToolset.Dtf.Compression 4.0.0-preview.1-build.88 was n
ot found. An approximate best match of WixToolset.Dtf.Compression 4.0.0-preview.1-build.89 was resolved.
C:\Solutions\TemporarySolutions\wix4\src\internal\WixBuildTools.TestSupport\WixBuildTools.TestSupport.csproj : warning NU1603: WixBuildTools.
TestSupport depends on WixToolset.Dtf.Compression.Cab (>= 4.0.0-preview.1-build.88) but WixToolset.Dtf.Compression.Cab 4.0.0-preview.1-build.
88 was not found. An approximate best match of WixToolset.Dtf.Compression.Cab 4.0.0-preview.1-build.89 was resolved.
C:\Solutions\TemporarySolutions\wix4\src\internal\WixBuildTools.TestSupport\WixBuildTools.TestSupport.csproj : warning NU1603: WixBuildTools.
TestSupport depends on WixToolset.Dtf.WindowsInstaller (>= 4.0.0-preview.1-build.88) but WixToolset.Dtf.WindowsInstaller 4.0.0-preview.1-buil
d.88 was not found. An approximate best match of WixToolset.Dtf.WindowsInstaller 4.0.0-preview.1-build.89 was resolved.
```
But then receive an error of:
```
       "C:\Solutions\TemporarySolutions\wix4\src\libs\libs.proj" (default target) (1) ->
       "C:\Solutions\TemporarySolutions\wix4\src\libs\dutil\dutil.proj" (default target) (2) ->
       "C:\Solutions\TemporarySolutions\wix4\src\libs\dutil\WixToolset.DUtil\dutil.vcxproj" (default target) (3) ->
       (ClCompile target) ->
         C:\Program Files (x86)\Windows Kits\10\Include\10.0.19041.0\um\windows.h(167): fatal error C1083: Cannot open include file: 'excpt.h
       ': No such file or directory [C:\Solutions\TemporarySolutions\wix4\src\libs\dutil\WixToolset.DUtil\dutil.vcxproj]


       "C:\Solutions\TemporarySolutions\wix4\src\libs\libs.proj" (default target) (1) ->
       "C:\Solutions\TemporarySolutions\wix4\src\libs\wcautil\wcautil.proj" (default target) (4) ->
       "C:\Solutions\TemporarySolutions\wix4\src\libs\wcautil\WixToolset.WcaUtil\wcautil.vcxproj" (default target) (5) ->
         C:\Program Files (x86)\Windows Kits\10\Include\10.0.19041.0\um\windows.h(167): fatal error C1083: Cannot open include file: 'excpt.h
       ': No such file or directory [C:\Solutions\TemporarySolutions\wix4\src\libs\wcautil\WixToolset.WcaUtil\wcautil.vcxproj]


       "C:\Solutions\TemporarySolutions\wix4\src\libs\libs.proj" (default target) (1) ->
       "C:\Solutions\TemporarySolutions\wix4\src\libs\wcautil\wcautil.proj" (default target) (4) ->
       "C:\Solutions\TemporarySolutions\wix4\src\libs\wcautil\WixToolset.WcaUtil\wcautil.vcxproj" (default target) (5:2) ->
         C:\Program Files (x86)\Windows Kits\10\Include\10.0.19041.0\um\windows.h(167): fatal error C1083: Cannot open include file: 'excpt.h
       ': No such file or directory [C:\Solutions\TemporarySolutions\wix4\src\libs\wcautil\WixToolset.WcaUtil\wcautil.vcxproj]


       "C:\Solutions\TemporarySolutions\wix4\src\libs\libs.proj" (default target) (1) ->
       "C:\Solutions\TemporarySolutions\wix4\src\libs\dutil\dutil.proj" (default target) (2) ->
       "C:\Solutions\TemporarySolutions\wix4\src\libs\dutil\WixToolset.DUtil\dutil.vcxproj" (default target) (3:2) ->
         C:\Program Files (x86)\Windows Kits\10\Include\10.0.19041.0\um\windows.h(167): fatal error C1083: Cannot open include file: 'excpt.h
       ': No such file or directory [C:\Solutions\TemporarySolutions\wix4\src\libs\dutil\WixToolset.DUtil\dutil.vcxproj]

    0 Warning(s)
    4 Error(s)

Time Elapsed 00:00:55.33
```

I presume that this is an issue with the windows SDK (I have the default 19041 installed). Or do I need the ATL's for 141 & 142 installed?

**Comment by robmen** [comment]
Others are better at diagnosing Windows SDK issues (and that looks like your problem) but I'd note that you'll want the very latest changes (as of this morning) to be able to successfully build using VS2022 (there are a few fixes).

**Reply by MarkStega** [reply]
@robmen Which version(s) of the Windows SDK do you have installed? (I am presuming that it is a Win 10 SDK since that is what VS2022 installs when you pick desktop C++ development)

**Reply by robmen** [reply]
@barnson and @rseanhall had "workload" definitions for VS2019 and one of them (I think Sean) tried to minimize his. Perhaps they can share and we can start on one for VS2022.

**Reply by barnson** [reply]
`excpt.h` is in the VC++ tree. My guess is that you're missing one the `v14x` build tools: You need `MSVC` build tools for both ARM64 and x86/x64, across v141, v142, and v143.

(For now. I'm testing whether we can get away with building only with v143.)

**Comment by MarkStega** [comment]
I have all three of the 141/142/143 build tools installed.

I found the files; There are 3 copies in "c:\program files\microsoft visual studio\2022\community\vc\tools\XXXX\include" where XXX is 14.16.2703, 14.29.30133, 14.30.3075

Since the error is "C:\Program Files (x86)\Windows Kits\10\Include\10.0.19041.0\um\windows.h(167): fatal error C1083: Cannot open include file: 'excpt.h'" it appears that the issue is one of include paths.

Do I need to set include paths?

I am attempting the build from a VS2022 x86 command prompt. Is that not correct?

**Reply by rseanhall** [reply]
We are currently building with v140, 141 and 142.

No, that is not correct. You should use the generic Developer Command Prompt for VS2022 (I hope that exists, it definitely did for VS2019).

**Comment by MarkStega** [comment]
Sorry, I have V140 build tools installed also; I tried the generic developer prompt with the same result. I was incorrectly interpreting the instruction from the V3 build document to start a 32 bit command prompt.

**Reply by rseanhall** [reply]
Here's my configuration for VS2022. I tried to install only things that were required for WiX. I went with the Windows 11 SDK since that's what is on the CI machines. This did require my changes [here](https://github.com/wixtoolset/wix4/pull/56), but that probably could have been worked around by installing the .NET 4.0 targeting pack.

[vs2022_wix.vsconfig.txt](https://github.com/wixtoolset/issues/files/7664009/vs2022_wix.vsconfig.txt)

**Comment by MarkStega** [comment]
As of today (12/7) when I try to build with the head commit (after the @rseanhall PR was merged) I still get the errors attempting to include excpt.h

However, if I check out the @barnson branch 'bob/vs2022' the build goes on for about 8 minutes, generates 17 nupkg's and 1 snupkg in artifacts.

There were test errors of

```
WixRunner failed with exit code: 103
   Output: C:\Solutions\TemporarySolutions\wix4\build\wix\Release\netcoreapp3.1\TestData\MultiFileCompressed\PackageComponents.wxs(6) : error TEST0103: The system cannot find the file 'C:\WINDOWS\Notepad.exe' with type 'File'. The following paths were checked: C:\WINDOWS\Notepad.exe
```
which is clearly an environment error, not a true test failure. And after figuring out how to install Notepad, the build runs to completion the following errors:
```
  C:\Solutions\TemporarySolutions\wix4\src\ext\Util\wixlib\util.wixproj : error : Unable to find package WixToolset.Sdk with version (= 4.0.0
-preview.1-build.88)
C:\Solutions\TemporarySolutions\wix4\src\ext\Util\wixlib\util.wixproj : error :   - Found 1 version(s) in build [ Nearest version: 4.0.0-prev
iew.1-build.96 ]
C:\Solutions\TemporarySolutions\wix4\src\ext\Util\wixlib\util.wixproj : error :   - Found 0 version(s) in nuget
  C:\Solutions\TemporarySolutions\wix4\src\ext\Util\wixlib\util.wixproj : error : C:\Program Files\dotnet\sdk\6.0.100\Sdks\WixToolset.Sdk\Sdk
 not found. Check that a recent enough .NET SDK is installed and/or increase the version specified in global.json.

**Reply by rseanhall** [reply]
The build has state, and there is no official way to clean everything. It seems people are using git commands to do that, but you need to be careful that those only delete files you're comfortable with deleting.

**Reply by barnson** [reply]
`git clean -ndx` followed by `git clean -fdx` and deep prayer. Also, nuke your NuGet package cache of everything WiX related. Yeah, it sucks.

**Comment by barnson** [comment]
Odd. Maybe the centralization of `PlatformToolset` had something to do with it...?

**Reply by rseanhall** [reply]
Not really. With your branch, only the v143 toolset is required. This just confirms that something is wrong with the earlier toolsets on that machine.

**Comment by MarkStega** [comment]
**Windows**
  **Nuget components**
    Nuget.exe (needs to be placed in a directory that is included in the path environment)
  **Optional components**
    Notepad.exe (used for some tests)
    
**VS2022 17.0.2**
  **Workloads**
    ASP.NET and web development
    Desktop development with C++
  **Individual components (beyond those added by the workloads)**
    MSVC v140 - VS2015 C++ build tools (v14.00)
    MSVC v141 - VS2017 C++ ARM64 build tools (v14.16) 
    MSVC v141 - VS2017 C++ x64/x86 build tools (v14.16) 
    MSVC v142 - VS2019 C++ ARM64 build tools (v14.29-16.11) 
    MSVC v142 - VS2019 C++ x64/x86 build tools (v14.29-16.11) 
    MSVC v143 - VS2022 C++ ARM64 build tools (latest) 
    MSVC v143 - VS2022 C++ x64/x86 build tools (latest) 

@rseanhall All of the earlier toolsets were installed from the VS2022 installer

**Reply by barnson** [reply]
We discussed on wix-devs that we need to choose the lowest version of the toolset for libraries and the highest for linked code so the v143-only approach won't work. BTW, ARM isn't needed (though ARM64 is).

**Reply by MarkStega** [reply]
@barnson - I updated the comment above per your comment about ARM; I am presuming that the need for 141/142 will go away at some point.

**Reply by rseanhall** [reply]
I spent some extra time yesterday trying to clean my machine and minimize what VS2022 installed. I'm not really interested in spending any more time diagnosing this. It works everywhere else. I recommend uninstalling and then reinstalling VS2022 using the configuration I provided.

**Reply by MarkStega** [reply]
@rseanhall Thanks -- Unfortunately, I have no idea what to do with the list of components that you supplied. I am going to wait until v4 has some proper developer documentation before trying again, it is just too much an uphill battle to try to figure out a working dev environment in the current state,

**Reply by rseanhall** [reply]
https://docs.microsoft.com/en-us/visualstudio/install/import-export-installation-configurations?view=vs-2022

**Comment by MarkStega** [comment]
Sean,

I actually figured out the procedure about an hour after I wrote my comment. Several of the components aren't available in VS2022 Community edition:

![2021-12-08_13-20-46](https://user-images.githubusercontent.com/634561/145262996-27df709a-2307-4624-b726-0a29c9f77da3.png)
.

**Reply by robmen** [reply]
Fortunately, you should not need those components to build the WiX Toolset. Hopefully, the rest of the missing Windows SDK bits *will* get turned on though so you can build native code.

The error code about `excpt.h` doesn't really provide much information to go on to diagnose what is unique about your machine to create that failure. So here's to hoping.

**Comment by MarkStega** [comment]
I first just modified my VS2022 with the import file -- Same 4 errors with excpt.h

I then totally removed VS2022 and reinstalled, again using the import file -- Same 4 errors...

**Reply by barnson** [reply]
Always with Community edition? It _should_ work but I don't think anyone working on WiX v4 right now is using it.

**Reply by rseanhall** [reply]
Hmm. You should probably create a bug with Visual Studio. It looks like Appveyor's VS2022 image is using Community edition though, and it built fine there with the changes I recently pushed.

**Reply by barnson** [reply]
Oh yeah, I forgot about the CI build.

**Comment by MarkStega** [comment]
Two quick questions:

* Is the appveyor build accessible publicly & if so, at what URL?

* Is Bob's (@barnson) branch likely to be merged as the official build? The reason I ask is that I can build that branch with no issues and I'd like to stop messing with trying to understand the build and the issues that I am having with the head commit from the master branch.

**Reply by rseanhall** [reply]
* We've stopped using Appveyor for our builds, but I haven't turned off mine yet - https://ci.appveyor.com/project/SeanHall/wix4/history.

* Not exactly, we decided to set the baseline for the static libs to v141. That branch is a work in progress.

**Reply by barnson** [reply]
`bob/vs2022` is up to date with the `v141`/`v143` plans but we're currently blocked getting a PR build (or official build) because the hosted GitHub image doesn't contain the `v141` toolset. Not sure how/when we solve that problem.

**Reply by MarkStega** [reply]
@rseanhall @barnson Thanks for the replies, that helps

Bob, Couldn't you download the old VS CLI Tools 141 package (https://visualstudio.microsoft.com/vs/older-downloads/), put the installer image in the WiX4 repository, and then execute the installer as an early step in the build yaml? Not ideal, but it should work.

[Edit] Arghh -- I decided to try this and the download is fine but it invokes the Visual Studio installer which certainly won't work in GitHub actions. Perhaps just adding the installed tools to the repository & updating a path variable? Now I know why you said that the problem wasn't solved :-)

**Reply by barnson** [reply]
Rob's experimenting with that in https://github.com/wixtoolset/wix4/pull/65.

**Reply by robmen** [reply]
Adds a chunk of time to the build, so haven't decided if we'll take the change like this now or not. I'd also like to see Bob's the `v141` work completed as well as it'll trim the amount we have to have installed and have to build.

**Comment by MarkStega** [comment, marked as answer]
This was resolved with the build enhancements that support building with VS2022
